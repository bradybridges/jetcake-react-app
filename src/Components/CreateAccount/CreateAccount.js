import React, { Component } from 'react';
import './CreateAccount.scss';
import * as firebase from 'firebase';
import 'firebase/auth';
import 'firebase/storage';
import 'firebase/firestore';

export default class CreateAccount extends Component {
  state = {
    email: "",
    phone: "",
    dob: "",
    profile: "",
    address: "",
    securityOne: "",
    securityTwo: "",
    securityThree: "",
    password: "",
    passwordConfirm: "",
    error: null,
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  }

  verifyPasswords = () => {
    const { password, passwordConfirm } = this.state;
    if((password === passwordConfirm) && (password !== "" && passwordConfirm !== "")) return true;
    return false;
  }

  verifyFormCompletion = () => {
    const { 
      email, 
      password, 
      passwordConfirm, 
      address, 
      phone, 
      profile, 
      dob, 
      securityOne, 
      securityTwo, 
      securityThree
    } = this.state;
    const requiredInfo = [
      email, 
      password, 
      passwordConfirm, 
      address, 
      phone, 
      profile, 
      dob, 
      securityOne, 
      securityTwo, 
      securityThree
    ];
    let completed = true;
    requiredInfo.forEach((field) => {
      if(field === "") {
        completed = false;
      }
    });
    return completed;
  }

  handleCreateAccount = async (e) => {
    e.preventDefault();
    const { toggleShowCreateAccount } = this.props;
    const { email, password } = this.state;
    try {
      if(!this.verifyPasswords()) {
        this.setState({ error: "Passwords don't match" });
        return;
      } else if(!this.verifyFormCompletion()) {
        this.setState({ error: "Please fill in all fields" });
        return;
      }
      const newUser = await firebase.auth().createUserWithEmailAndPassword(email, password);
      await this.handleProfilePictureUpload(newUser);
      await this.handleAddProfile();
      toggleShowCreateAccount();
    } catch(error) {
      this.setState({ error: error.message });
    }
  }

  handleAddProfile = async () => {
    const { 
      email, 
      address, 
      phone, 
      dob, 
      securityOne, 
      securityTwo, 
      securityThree
    } = this.state;
    const newProfile ={
      email,
      address,
      phone,
      dob,
      securityOne,
      securityTwo,
      securityThree,
    }
    await firebase.firestore().collection('profiles').add(newProfile);
  }

  handleImgChange = (e) => {
    if(e.target.files[0]) {
      this.setState ({ profile: e.target.files[0] });
    }
  }

  handleProfilePictureUpload = async (newUser) => {
    const { email } = newUser.user;
    const { profile } = this.state;
    try {
      await firebase.storage().ref(`${email}/profile.jpg`).put(profile);
      const url = await firebase.storage().ref(email).child('profile.jpg').getDownloadURL();
      console.log(url);
      this.updateProfilePicture(url);
    } catch(error) {
      console.error('Upload Error', error);
    }
  }

  updateProfilePicture = async (photoURL) => {
    const user = await firebase.auth().currentUser;
    await user.updateProfile({
      photoURL
    });
  }

  render() {
    const { toggleShowCreateAccount } = this.props;
    return (
      <form id="create-account-form" onSubmit={this.handleCreateAccount}>
        {this.state.error && <h3>{this.state.error}</h3>}
        <label>Email</label>
        <input name="email" onChange={this.handleChange} type="email" value={this.state.email}/>
        <label>Password</label>
        <input type="password" onChange={this.handleChange} name="password" value={this.state.password}/>
        <label>Confirm Password</label>
        <input type="password" onChange={this.handleChange} name="passwordConfirm" value={this.state.passwordConfirm}/>
        <label>Phone Number</label>
        <input name="phone" onChange={this.handleChange} type="text" value={this.state.phone}/>
        <label>Date of Birth</label>
        <input name="dob" onChange={this.handleChange} type="date" value={this.state.dob}/>
        <label>Profile Picture</label>
        <input name="profile" onChange={this.handleImgChange} type="file" />
        <label>Address</label>
        <input name="address" onChange={this.handleChange} type="text" value={this.state.address}/>
        <section id="security-questions-container">
          <span>Mothers Maiden Name?</span>
          <input name="securityOne" onChange={this.handleChange} type="text" value={this.state.securityOne}/>
          <span>First Pet's Name?</span>
          <input name="securityTwo" onChange={this.handleChange} type="text" value={this.state.securityTwo}/>
          <span>Town where you grew up?</span>
          <input name="securityThree" onChange={this.handleChange} type="text" value={this.state.securityThree}/>
        </section>
        <section id="submission-container">
          <button type="submit">Submit</button>
          <button onClick={toggleShowCreateAccount}>Cancel</button>
        </section>
      </form>
    )
  }
}