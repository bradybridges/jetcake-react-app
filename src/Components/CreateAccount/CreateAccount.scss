#create-account-form {
  display: flex;
  flex-direction: column;
  width: 400px;
  position: absolute;
  top: 20%;
  left: 50%;
  margin-left: -200px;
  justify-content: center;
  align-items: center;
  padding: 10px;
  border-radius: 4px;
  box-shadow: 0px 0px 20px 0px #ffffff2e;
  background-color: rgba(0, 0, 0, 0.95);
}

#security-questions-container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.error {
  color: white;
}
