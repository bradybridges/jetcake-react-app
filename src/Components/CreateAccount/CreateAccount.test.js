import React from 'react';
import { shallow } from 'enzyme';
import CreateAccount from './CreateAccount';

describe('CreateAccount', () => {
  let wrapper;
  const mockToggleShowCreateAccount = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<CreateAccount toggleShowCreateAccount={mockToggleShowCreateAccount} />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('handleChange should change corresponding property in state with a given value', () => {
    const mockEmailEvent = {
      target: {
        name: 'email',
        value: 'newEmailHere@yahoo.com',
      },
    };
    const mockPhoneEvent = {
      target: {
        name: 'phone',
        value: '0000000000',
      },
    };
    wrapper.instance().handleChange(mockEmailEvent);
    expect(wrapper.state('email')).toEqual('newEmailHere@yahoo.com');
    wrapper.instance().handleChange(mockPhoneEvent);
    expect(wrapper.state('phone')).toEqual('0000000000');
  });

  it('verifyPasswords should make sure the passwords provided are the same', () => {
    wrapper.setState({ password: 'password', passwordConfirm: 'password' });
    let result = wrapper.instance().verifyPasswords();
    expect(result).toEqual(true);
    wrapper.setState({ passwordConfirm: 'not matching' });
    result = wrapper.instance().verifyPasswords();
    expect(result).toEqual(false);
  });

  it('verifyFormCompletion should make sure all fields have input', () => {
    let result = wrapper.instance().verifyFormCompletion();
    expect(result).toEqual(false);
    wrapper.setState({
      phone: '1231231234',
      dob: '2000-01-01',
      profile: '<img>',
      address: '123 test ct',
      securityOne: 'answer1',
      securityTwo: 'answer2',
      securityThree: 'answer3',
      password: 'password',
      passwordConfirm: 'password',
    });
    result = wrapper.instance().verifyFormCompletion();
    expect(result).toEqual(false);
    wrapper.setState({ email: 'testEmail@yahoo.com' });
    result = wrapper.instance().verifyFormCompletion();
    expect(result).toEqual(true);
  });

  it('handleImgChange should update state of new image added by user', () => {
    const mockEvent = {
      target: {
        files: ['this is an image'],
      },
    };
    expect(wrapper.state('profile')).toEqual('');
    wrapper.instance().handleImgChange(mockEvent);
    expect(wrapper.state('profile')).toEqual('this is an image');
  });
});
