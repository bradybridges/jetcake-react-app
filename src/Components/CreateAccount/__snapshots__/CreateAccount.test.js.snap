// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccount should match snapshot 1`] = `
<form
  id="create-account-form"
  onSubmit={[Function]}
>
  <h3
    id="create-account-header"
  >
    Create An Account
  </h3>
  <label
    className="input-label"
  >
    Email
  </label>
  <input
    className="input-field"
    name="email"
    onChange={[Function]}
    type="email"
    value=""
  />
  <label
    className="input-label"
  >
    Password
  </label>
  <input
    className="input-field"
    name="password"
    onChange={[Function]}
    type="password"
    value=""
  />
  <label
    className="input-label"
  >
    Confirm Password
  </label>
  <input
    className="input-field"
    name="passwordConfirm"
    onChange={[Function]}
    type="password"
    value=""
  />
  <label
    className="input-label"
  >
    Phone Number
  </label>
  <input
    className="input-field"
    name="phone"
    onChange={[Function]}
    type="text"
    value=""
  />
  <label
    className="input-label"
  >
    Date of Birth
  </label>
  <input
    className="input-field"
    name="dob"
    onChange={[Function]}
    type="date"
    value=""
  />
  <label
    className="input-label"
  >
    Profile Picture
  </label>
  <input
    className="input-field"
    name="profile"
    onChange={[Function]}
    type="file"
  />
  <label
    className="input-label"
  >
    Address
  </label>
  <input
    className="input-field"
    name="address"
    onChange={[Function]}
    type="text"
    value=""
  />
  <section
    id="security-questions-container"
  >
    <label
      className="input-label"
    >
      Mothers Maiden Name?
    </label>
    <input
      className="input-field"
      name="securityOne"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      className="input-label"
    >
      First Pet's Name?
    </label>
    <input
      className="input-field"
      name="securityTwo"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      className="input-label"
    >
      Town where you grew up?
    </label>
    <input
      className="input-field"
      name="securityThree"
      onChange={[Function]}
      type="text"
      value=""
    />
  </section>
  <section
    className="form-btn-container"
  >
    <button
      className="form-btn"
      type="submit"
    >
      Submit
    </button>
    <button
      className="form-btn"
      onClick={[Function]}
    >
      Cancel
    </button>
  </section>
</form>
`;
